hcvault/keysource.go

```diff
@@ -0,0 +1,274 @@
+	}
+	encryptedKey, ok := encrypted.(string)
+	if !ok {
+		return fmt.Errorf("the ciphertext cannot be casted to string")
+	}
+	key.EncryptedKey = encryptedKey
+	return nil
+}
+
+// EncryptIfNeeded encrypts the provided sops' data key and encrypts it if it hasn't been encrypted yet
+func (key *MasterKey) EncryptIfNeeded(dataKey []byte) error {
+	if key.EncryptedKey == "" {
+		return key.Encrypt(dataKey)
+	}
+	return nil
+}
+
+// Decrypt decrypts the EncryptedKey field with Vault Transit and returns the result.
+func (key *MasterKey) Decrypt() ([]byte, error) {
+	fullPath := path.Join(key.BackendPath, "decrypt", key.KeyName)
+	cli, err := vaultClient(key.VaultAddress)
+	if err != nil {
+		return nil, err
+	}
+	payload := make(map[string]interface{})
+	payload["ciphertext"] = key.EncryptedKey
+	raw, err := cli.Logical().Write(fullPath, payload)
+	if err != nil {
+		log.WithField("Path", fullPath).Info("Encryption failed")
+		return nil, err
+	}
+	if raw == nil || raw.Data == nil {
+		return nil, fmt.Errorf("The transit backend %s is empty", fullPath)
+	}
+	decrypted, ok := raw.Data["plaintext"]
+	if ok != true {
+		return nil, fmt.Errorf("there's no decrypted data")
+	}
+	dataKey, ok := decrypted.(string)
+	if ok != true {
+		return nil, fmt.Errorf("the plaintest cannot be casted to string")
+	}
+	result, err := base64.StdEncoding.DecodeString(dataKey)
+	if err != nil {
+		return nil, fmt.Errorf("Couldn't decode base64 plaintext")
+	}
+	return result, nil
+}
+
+// NeedsRotation returns whether the data key needs to be rotated or not.
+// This is simply copied from GCPKMS
+// TODO: handle key rotation on vault side
+func (key *MasterKey) NeedsRotation() bool {
+	//TODO: manage rewrapping https://www.vaultproject.io/api/secret/transit/index.html#rewrap-data
+	return time.Since(key.CreationDate) > (time.Hour * 24 * 30 * 6)
+}
+
+// ToString converts the key to a string representation
+func (key *MasterKey) ToString() string {
+	return fmt.Sprintf("%s/v1/%s/keys/%s", key.VaultAddress, key.BackendPath, key.KeyName)
+}
+
+func (key *MasterKey) createVaultTransitAndKey() error {
+	cli, err := vaultClient(key.VaultAddress)
+	if err != nil {
+		return err
+	}
+	if err != nil {
+		return fmt.Errorf("Cannot create Vault Client: %v", err)
+	}
+	err = cli.Sys().Mount(key.BackendPath, &api.MountInput{
+		Type:        "transit",
+		Description: "backend transit used by SOPS",
+	})
+	if err != nil {
+		return err
+	}
+	path := path.Join(key.BackendPath, "keys", key.KeyName)
+	payload := make(map[string]interface{})
+	payload["type"] = "rsa-4096"
+	_, err = cli.Logical().Write(path, payload)
+	if err != nil {
+		return err
+	}
+	_, err = cli.Logical().Read(path)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// ToMap converts the MasterKey to a map for serialization purposes
+func (key MasterKey) ToMap() map[string]interface{} {
+	out := make(map[string]interface{})
+	out["vault_address"] = key.VaultAddress
+	out["keyname"] = key.KeyName
```

Comment:
just for uniformity with config snake case variant
```suggestion
	out["key_name"] = key.KeyName
```